### PR TITLE
Implemented Discovery Server for local participant

### PR DESCRIPTION
**Pull request by rsanchez15**

### CODE REVIEW

File: src/cpp/DataBrokerConfiguration.cpp [hunk 1]
```diff
@@ -254,6 +260,37 @@ bool DataBrokerConfiguration::load_configuration_file(
             logInfo(DATABROKER_CONFIGURATION, "Using internal domain " << configuration.local_configuration.domain);
         }
 
+        // LOCAL DISCOVERY
+        if (config_node["local-discovery"])
+        {
+            if (config_node["local-discovery"]["kind"])
+            {
+                if (config_node["local-discovery"]["kind"].as<std::string>() == "discovery-server")
```
Comment (jparisu) [inline src/cpp/DataBrokerConfiguration.cpp:268]: try not to rely in strings that set the type

File: src/cpp/DataBrokerConfiguration.cpp [hunk 2]
```diff
@@ -254,6 +260,37 @@ bool DataBrokerConfiguration::load_configuration_file(
...
+                        {
+                            new_address.port = address["port"].as<uint32_t>();
+                        }
+                        configuration.local_configuration.discovery_config.listening_addresses.push_back(new_address);
+                    }
+
+                    configuration.local_configuration.discovery_config.server_guid =
+                            Address::ros_discovery_server_guid();
+                }
+            }
```
Comment (jparisu) [inline src/cpp/DataBrokerConfiguration.cpp:291]: If guid will not be set here, do not store it in the configuration and use always the default one. I would suggest to give the option to set it here and use `ROS_DISCOVERY_SERVER_GUID` as default (to make it more configurable)(use WAN configuration to copy `listening-addresses` configuration parser)

**Review by jparisu — CHANGES_REQUESTED** [review]

File: src/cpp/DataBrokerWANParticipant.cpp [hunk 3]
```diff
@@ -138,9 +138,9 @@ eprosima::fastdds::dds::DomainParticipantQos DataBrokerWANParticipant::participa
     }
 
     // TODO decide the discovery server configuration
-    pqos.wire_protocol().builtin.discovery_config.leaseDuration = fastrtps::c_TimeInfinite;
-    pqos.wire_protocol().builtin.discovery_config.leaseDuration_announcementperiod =
-            fastrtps::Duration_t(2, 0);
+    // pqos.wire_protocol().builtin.discovery_config.leaseDuration = fastrtps::c_TimeInfinite;
```
[the remark this answers is not in the document]
Comment (rsanchez15) [reply src/cpp/DataBrokerWANParticipant.cpp:141]: Let keep the default values:
 - Lease duration 20s
 - Announcement period 3s

File: src/cpp/DataBrokerConfiguration.cpp (hunk 2 above)
Comment (rsanchez15) [reply src/cpp/DataBrokerConfiguration.cpp:291]: Not for the PoC version

File: include/databroker/DataBrokerConfiguration.hpp [hunk 4]
```diff
@@ -45,8 +46,13 @@ struct DataBrokerParticipantConfiguration
 struct DataBrokerLocalParticipantConfiguration : public DataBrokerParticipantConfiguration
 {
     bool ros;
+    bool discovery_server;
+    eprosima::fastrtps::rtps::DiscoveryProtocol discovery_protocol;
+    std::vector<Address> listening_addresses;
+    eprosima::fastrtps::rtps::GuidPrefix_t server_guid;
```
Comment (jparisu) [inline include/databroker/DataBrokerConfiguration.hpp:52]:
```suggestion
    std::vector<Address> listening_addresses;
```
I would erase those as they are not parameters, but are always the same value depending on `discovery_server`

File: src/cpp/DataBrokerConfiguration.cpp [hunk 5]
```diff
@@ -254,6 +261,47 @@ bool DataBrokerConfiguration::load_configuration_file(
...
+                            new_address.ip = address["ip"].as<std::string>();
+                        }
+                        if (address["port"])
+                        {
+                            new_address.port = address["port"].as<uint32_t>();
+                        }
+                        configuration.local_configuration.listening_addresses.push_back(new_address);
+                    }
+
+                    if (empty_listening_addresses)
```
Comment (jparisu) [inline src/cpp/DataBrokerConfiguration.cpp:293]:
```suggestion
                    if (configuration.local_configuration.listening_addresses.empty())
```

File: src/cpp/DataBrokerLocalParticipant.cpp [hunk 6]
```diff
@@ -30,6 +33,45 @@ DataBrokerLocalParticipant::DataBrokerLocalParticipant(
...
+eprosima::fastdds::dds::DomainParticipantQos DataBrokerLocalParticipant::participant_qos()
+{
+    eprosima::fastdds::dds::DomainParticipantQos pqos = DataBrokerParticipant::participant_qos();
+
+    if (configuration_.discovery_protocol ==
+            fastrtps::rtps::DiscoveryProtocol::SERVER)
+    {
+        // Set this participant as a SERVER
+        pqos.wire_protocol().builtin.discovery_config.discoveryProtocol =
+                fastrtps::rtps::DiscoveryProtocol::SERVER;
```
Comment (jparisu) [inline src/cpp/DataBrokerLocalParticipant.cpp:57]: You have the variable in the configuration and you don't even use it :crying_cat_face:

**Review by jparisu — CHANGES_REQUESTED** [review]
Add this new feature to the example configuration file

**Review by jparisu — APPROVED** [review]
LGTM

**Pull request merged**